src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -82,51 +79,29 @@
                                           (filter dir-create?)
                                           (map #(.toPath (:path %)))))))
 
-(defn reset-key-and-unwatch-deleted!
-  [watch-key]
-  ;; reset key and remove from set if directory no longer accessible
-  (let [valid? (.reset watch-key)]
-    (when (not valid?)
-      ;; TODO: not sure if we need to do anything else here now that we're not retaining a map of
-      ;; these keys.  The Oracle example code makes it seem like maybe you do, but I can't tell what.
-      (log/info (trs "Removing watched directory {0} because it was deleted" (.watchable watch-key))))))
-
-;; This function (and other helper functions in this namespace)
-;; are heavily influenced by WatchDir.java from Java's official documentation:
-;; https://docs.oracle.com/javase/tutorial/essential/io/notification.html
-(schema/defn handle-watch-events!
-  [watchers :- [(schema/protocol Watcher)]
+(schema/defn watch!
+  "Returns a future representing processing events for the passed in watcher.
+  The future will swallow errors, loop producing side-effects until stopped,
+  and is expected to coordinate with other functions via the stopped? atom."
+  [watcher :- (schema/protocol Watcher)
+   stopped? :- Atom
    shutdown-on-error :- IFn]
-
-  ;; If we were going to add any additional "debouncing" here I think it
-  ;; would look something like this:
-  ;;
-  ;(let [mas? (atom true)
-  ;      watch-keys (atom [])]
-  ;  (while @mas?
-  ;    (if-let [watch-key (.poll watch-service)]
-  ;      (swap! watch-keys conj watch-key)
-  ;      (reset! mas? false))))
-  ;;
-  ;; (yes, I know that code is ugly as sin, and actually, as Chris points out,
-  ;; perhaps what we'd want to do is something like write a small implementation
-  ;; of Java's Iterator interface that wrapped the call to .poll,
-  ;; and then you could use Clojure's iterator-seq.)
-  ;;
-  ;; ... The idea being that we'd collect all of the available watch keys
-  ;; first, and then process them as a batch.  I think this could result
-  ;; in fewer callback invocations for nested watch keys (i.e. files
-  ;; nested under the root watched path) but the extent to which that is
-  ;; true is likely platform-dependent.
-  ;;
-  ;; https://tickets.puppetlabs.com/browse/PE-15621
-
-  (doseq [watcher watchers]
-    (when-let [watch-key (.poll (:watch-service watcher))]
-      (let [orig-events (.pollEvents watch-key)
+  (future
+    (while (not @stopped?)
```

Comment:
It looks like `stopped?` is currently only read and never written to.  I think you need to add something like `(reset! stopped? true)` inside the `catch` expression a few lines below.

Also, it looks like `stopped?` could be a local binding inside this function (in a `let` expression immediately above) instead of part of the TK service context.